### ISSUE TITLE
PG::ProtocolViolation

### ISSUE THREAD

**Issue by hx** [rails/rails#20236]
I've found that a certain relation doesn't take kindly to further scoping with a `where` clause.

Here's the test case:

```ruby
system 'ruby -v'

gem 'edge', '0.4.1'
gem 'activerecord', '4.2.1'
gem 'pg', '0.18.2'

require 'edge'

ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  host: 'localhost',
  port: 5432,
  username: 'test',
  password: 'test'
)

ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveSupport::LogSubscriber.colorize_logging = false

ActiveRecord::Schema.define do
  self.verbose = false
  create_table :things, force: true do |t|
    t.integer 'parent_id'
  end
end

class Thing < ActiveRecord::Base
  acts_as_forest
end

thing1 = Thing.create
thing2 = Thing.create parent_id: thing1.id

# This works:
puts Thing.where(id: thing1.id).with_descendants.pluck(:id)

# This doesn't:
puts Thing.where(id: thing1.id).with_descendants.find thing2.id
```

Here are the results:

```plaintext
ruby 2.2.1p85 (2015-02-26 revision 49769) [x86_64-darwin14]
D, [2015-05-21T10:01:00.299353 #89460] DEBUG -- :    (3.3ms)  DROP TABLE "things"
D, [2015-05-21T10:01:00.303992 #89460] DEBUG -- :    (4.3ms)  CREATE TABLE "things" ("id" serial primary key, "parent_id" integer) 
D, [2015-05-21T10:01:00.329463 #89460] DEBUG -- :    (0.2ms)  BEGIN
D, [2015-05-21T10:01:00.331791 #89460] DEBUG -- :   SQL (0.6ms)  INSERT INTO "things" DEFAULT VALUES RETURNING "id"
D, [2015-05-21T10:01:00.332509 #89460] DEBUG -- :    (0.4ms)  COMMIT
D, [2015-05-21T10:01:00.332939 #89460] DEBUG -- :    (0.1ms)  BEGIN
D, [2015-05-21T10:01:00.335323 #89460] DEBUG -- :   SQL (0.4ms)  INSERT INTO "things" ("parent_id") VALUES ($1) RETURNING "id"  [["parent_id", 1]]
D, [2015-05-21T10:01:00.335938 #89460] DEBUG -- :    (0.4ms)  COMMIT
D, [2015-05-21T10:01:00.339976 #89460] DEBUG -- :    (0.7ms)  SELECT "things"."id" FROM "things" WHERE (things.id in (WITH RECURSIVE "all_nodes" AS ( SELECT "things"."id", "things"."parent_id" FROM "things" WHERE "things"."id" = $1 UNION SELECT "things"."id", "things"."parent_id" FROM "things" INNER JOIN "all_nodes" ON "things"."parent_id" = "all_nodes"."id" ) SELECT "things"."id" FROM "all_nodes" INNER JOIN "things" ON "all_nodes"."id" = "things"."id"))  [["id", 1]]
1
2
D, [2015-05-21T10:01:00.342152 #89460] DEBUG -- :   Thing Load (0.3ms)  SELECT  "things".* FROM "things" WHERE (things.id in (WITH RECURSIVE "all_nodes" AS ( SELECT "things"."id", "things"."parent_id" FROM "things" WHERE "things"."id" = $1 UNION SELECT "things"."id", "things"."parent_id" FROM "things" INNER JOIN "all_nodes" ON "things"."parent_id" = "all_nodes"."id" ) SELECT "things"."id" FROM "all_nodes" INNER JOIN "things" ON "all_nodes"."id" = "things"."id")) AND "things"."id" = $1 LIMIT 1  [["id", 1], ["id", 2]]
E, [2015-05-21T10:01:00.342217 #89460] ERROR -- : PG::ProtocolViolation: ERROR:  bind message supplies 2 parameters, but prepared statement "a3" requires 1
: SELECT  "things".* FROM "things" WHERE (things.id in (WITH RECURSIVE "all_nodes" AS ( SELECT "things"."id", "things"."parent_id" FROM "things" WHERE "things"."id" = $1 UNION SELECT "things"."id", "things"."parent_id" FROM "things" INNER JOIN "all_nodes" ON "things"."parent_id" = "all_nodes"."id" ) SELECT "things"."id" FROM "all_nodes" INNER JOIN "things" ON "all_nodes"."id" = "things"."id")) AND "things"."id" = $1 LIMIT 1
/Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/postgresql_adapter.rb:602:in `exec_prepared': PG::ProtocolViolation: ERROR:  bind message supplies 2 parameters, but prepared statement "a3" requires 1 (ActiveRecord::StatementInvalid)
: SELECT  "things".* FROM "things" WHERE (things.id in (WITH RECURSIVE "all_nodes" AS ( SELECT "things"."id", "things"."parent_id" FROM "things" WHERE "things"."id" = $1 UNION SELECT "things"."id", "things"."parent_id" FROM "things" INNER JOIN "all_nodes" ON "things"."parent_id" = "all_nodes"."id" ) SELECT "things"."id" FROM "all_nodes" INNER JOIN "things" ON "all_nodes"."id" = "things"."id")) AND "things"."id" = $1 LIMIT 1
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/postgresql_adapter.rb:602:in `block in exec_cache'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/abstract_adapter.rb:473:in `block in log'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activesupport-4.2.1/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/abstract_adapter.rb:467:in `log'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/postgresql_adapter.rb:601:in `exec_cache'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/postgresql_adapter.rb:585:in `execute_and_clear'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/postgresql/database_statements.rb:160:in `exec_query'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/abstract/database_statements.rb:347:in `select'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/abstract/database_statements.rb:32:in `select_all'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/connection_adapters/abstract/query_cache.rb:70:in `select_all'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/querying.rb:39:in `find_by_sql'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/relation.rb:638:in `exec_queries'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/relation.rb:514:in `load'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/relation.rb:243:in `to_a'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/relation/finder_methods.rb:487:in `find_take'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/relation/finder_methods.rb:105:in `take'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/relation/finder_methods.rb:454:in `find_one'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/relation/finder_methods.rb:435:in `find_with_ids'
	from /Users/Neil/.rvm/gems/ruby-2.2.1/gems/activerecord-4.2.1/lib/active_record/relation/finder_methods.rb:71:in `find'
	from edge_testcase_pg.rb:38:in `<main>'
```

### TL;DR

The query has two bound values; `id: 1` and `id: 2`. There are two variables in the prepared SQL: `$1` and `$1`. It should be `$1` and `$2`, but instead, PG only wants one bound value, and complains that it has too many.

I prepared a similar test using an in-memory SQLite database, which doesn't suffer from the same issue, because it uses question marks instead of numbered variables in prepared statements. It's [here](https://gist.github.com/hx/cbe9a3e639832f63c9bb) if anyone wants it.

I'm happy to provide additional data should anyone concerned with the issue request it.

This may be related to <https://github.com/rails/rails/issues/15920>. It may also be related to another [gem](https://github.com/JackC/edge), though from my reviews of its code, it only uses AR methods (no SQL funny business).

**Comment by hx** [comment]
P.S. This issue arose after upgrading from Rails 3.2 to Rails 4.2.

**Comment by hx** [comment]
Closing after identifying a flaw in the <https://github.com/jackc/edge> gem.
